File: services/session-identity/test/property/oauth/pkce_property_test.exs
defmodule SessionIdentityCore.OAuth.PKCEPropertyTest do
  @moduledoc """
  Property tests for PKCE verification correctness.
  
  Property 2: PKCE Verification Correctness
  - For any valid code_verifier, computing S256 and verifying SHALL succeed
  - For any two different code_verifiers, cross-verification SHALL fail
  - For any string outside 43-128 chars, validation SHALL reject
  """

  use ExUnit.Case, async: true
  use ExUnitProperties

  alias SessionIdentityCore.OAuth.PKCE
  alias SessionIdentityCore.Test.Generators

  @iterations 100

  describe "Property 2: PKCE Verification Correctness" do
    property "valid code_verifier produces verifiable challenge" do
      check all(verifier <- Generators.code_verifier(), max_runs: @iterations) do
        challenge = PKCE.compute_s256_challenge(verifier)
        assert :ok = PKCE.verify(verifier, challenge, "S256")
      end
    end

    property "different verifiers produce different challenges" do
      check all(
              verifier1 <- Generators.code_verifier(),
              verifier2 <- Generators.code_verifier(),
              verifier1 != verifier2,
              max_runs: @iterations
            ) do
        challenge1 = PKCE.compute_s256_challenge(verifier1)
        challenge2 = PKCE.compute_s256_challenge(verifier2)
        assert challenge1 != challenge2
      end
    end

    property "cross-verification fails" do
      check all(
              verifier1 <- Generators.code_verifier(),
              verifier2 <- Generators.code_verifier(),
              verifier1 != verifier2,
              max_runs: @iterations
            ) do
        challenge1 = PKCE.compute_s256_challenge(verifier1)
        assert {:error, :invalid_code_verifier} = PKCE.verify(verifier2, challenge1, "S256")
      end
    end

    property "short verifiers are rejected" do
      check all(verifier <- Generators.short_code_verifier(), max_runs: @iterations) do
        assert {:error, :code_verifier_too_short} = PKCE.validate_code_verifier(verifier)
      end
    end

    property "long verifiers are rejected" do
      check all(verifier <- Generators.long_code_verifier(), max_runs: @iterations) do
        assert {:error, :code_verifier_too_long} = PKCE.validate_code_verifier(verifier)
      end
    end

    property "valid verifiers pass validation" do
      check all(verifier <- Generators.code_verifier(), max_runs: @iterations) do
        assert :ok = PKCE.validate_code_verifier(verifier)
      end
    end

    property "S256 challenge is exactly 43 characters" do
      check all(verifier <- Generators.code_verifier(), max_runs: @iterations) do
        challenge = PKCE.compute_s256_challenge(verifier)
        assert String.length(challenge) == 43
      end
    end

    property "plain method is always rejected" do
      check all(verifier <- Generators.code_verifier(), max_runs: @iterations) do
        challenge = PKCE.compute_s256_challenge(verifier)
        assert {:error, :plain_method_not_allowed} = PKCE.verify(verifier, challenge, "plain")
      end
    end
  end
end

File: sdk/typescript/tests/property/pkce.property.test.ts
/**
 * Property-based tests for PKCE implementation
 *
 * **Feature: typescript-sdk-modernization**
 * **Properties: 7, 8, 9**
 */

import { describe, it, expect } from 'vitest';
import * as fc from 'fast-check';
import { generatePKCE, verifyPKCE } from '../../src/pkce.js';

/**
 * **Property 7: PKCE S256 Method Enforcement**
 * For any PKCE challenge, codeChallengeMethod is always 'S256'.
 * **Validates: Requirements 5.1**
 */
describe('Property 7: PKCE S256 Method Enforcement', () => {
  it('always uses S256 method', async () => {
    await fc.assert(
      fc.asyncProperty(fc.constant(null), async () => {
        const challenge = await generatePKCE();
        expect(challenge.codeChallengeMethod).toBe('S256');
      }),
      { numRuns: 100 }
    );
  });
});

/**
 * **Property 8: PKCE Verifier Uniqueness**
 * For any two PKCE challenges, codeVerifier values are different.
 * **Validates: Requirements 5.2**
 */
describe('Property 8: PKCE Verifier Uniqueness', () => {
  it('generates unique verifiers', async () => {
    const verifiers = new Set<string>();
    const iterations = 100;

    for (let i = 0; i < iterations; i++) {
      const challenge = await generatePKCE();
      verifiers.add(challenge.codeVerifier);
    }

    // All verifiers should be unique
    expect(verifiers.size).toBe(iterations);
  });

  it('verifier length is within RFC 7636 bounds (43-128 chars)', async () => {
    await fc.assert(
      fc.asyncProperty(fc.constant(null), async () => {
        const challenge = await generatePKCE();
        expect(challenge.codeVerifier.length).toBeGreaterThanOrEqual(43);
        expect(challenge.codeVerifier.length).toBeLessThanOrEqual(128);
      }),
      { numRuns: 100 }
    );
  });

  it('verifier uses only valid base64url characters', async () => {
    const base64urlRegex = /^[A-Za-z0-9_-]+$/;

    await fc.assert(
      fc.asyncProperty(fc.constant(null), async () => {
        const challenge = await generatePKCE();
        expect(base64urlRegex.test(challenge.codeVerifier)).toBe(true);
      }),
      { numRuns: 100 }
    );
  });
});

/**
 * **Property 9: PKCE Round-Trip Verification**
 * For any PKCE challenge, verifying verifier against challenge returns true.
 * **Validates: Requirements 5.1, 5.2**
 */
describe('Property 9: PKCE Round-Trip Verification', () => {
  it('verifier matches its own challenge', async () => {
    await fc.assert(
      fc.asyncProperty(fc.constant(null), async () => {
        const challenge = await generatePKCE();
        const isValid = await verifyPKCE(
          challenge.codeVerifier,
          challenge.codeChallenge
        );
        expect(isValid).toBe(true);
      }),
      { numRuns: 100 }
    );
  });

  it('verifier does not match different challenge', async () => {
    await fc.assert(
      fc.asyncProperty(fc.constant(null), async () => {
        const challenge1 = await generatePKCE();
        const challenge2 = await generatePKCE();

        // Different verifiers should not match each other's challenges
        if (challenge1.codeVerifier !== challenge2.codeVerifier) {
          const isValid = await verifyPKCE(
            challenge1.codeVerifier,
            challenge2.codeChallenge
          );
          expect(isValid).toBe(false);
        }
      }),
      { numRuns: 100 }
    );
  });
});
